kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLCommandHandler.scala

```diff
@@ -0,0 +1,196 @@
+
+import io.netty.channel.{ChannelHandlerContext, SimpleChannelInboundHandler}
+import org.apache.hive.service.rpc.thrift.TProtocolVersion
+
+import org.apache.kyuubi.{KyuubiSQLException, Logging}
+import org.apache.kyuubi.operation.FetchOrientation
+import org.apache.kyuubi.operation.OperationState._
+import org.apache.kyuubi.server.mysql.MySQLCommandHandler._
+import org.apache.kyuubi.server.mysql.constant.MySQLCtxAttrKey._
+import org.apache.kyuubi.service.BackendService
+import org.apache.kyuubi.session.SessionHandle
+
+object MySQLCommandHandler {
+  val connIdCounter = new AtomicInteger
+  val connIdToSessHandle = new ConcurrentHashMap[Int, SessionHandle]
+}
+
+class MySQLCommandHandler(be: BackendService, execPool: ThreadPoolExecutor)
+  extends SimpleChannelInboundHandler[MySQLCommandPacket] with Logging {
+
+  private implicit val ec: ExecutionContextExecutor = ExecutionContext.fromExecutor(execPool)
+
+  @volatile private var closed: Boolean = false
+
+  override def channelInactive(ctx: ChannelHandlerContext): Unit = {
+    closeSession(ctx)
+    super.channelInactive(ctx)
+  }
+
+  // handle process exception, generally should send error packet
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    val connectionId = ctx.channel.attr(CONNECTION_ID).get
+    val errPacket = MySQLErrPacket(cause)
+    error(s"Connection: $connectionId, $errPacket")
+    if (ctx.channel.isActive) {
+      ctx.writeAndFlush(errPacket)
+    } else {
+      warn(s"Ignore error packet for inactivated connection: $connectionId")
+    }
+  }
+
+  override def channelRead0(ctx: ChannelHandlerContext, packet: MySQLCommandPacket): Unit = Future {
+    ensureSessionOpened(ctx)
+    packet match {
+      case pkt: MySQLComPingPacket => handlePing(ctx, pkt)
+      case pkt: MySQLComInitDbPacket => handleInitDb(ctx, pkt)
+      case pkt: MySQLComQuitPacket => handleQuit(ctx, pkt)
+      case pkt: MySQLComQueryPacket => handleQuery(ctx, pkt)
+      case bad => throw new UnsupportedOperationException(bad.getClass.getSimpleName)
+    }
+  } onComplete {
+    case Success(responsePackets) =>
+      responsePackets.foreach(ctx.channel.write)
+      ctx.channel.flush()
+    case Failure(cause) =>
+      exceptionCaught(ctx, cause)
+  }
+
+  def ensureSessionOpened(ctx: ChannelHandlerContext): Unit =
+    if (ctx.channel.attr(SESSION_HANDLE).get == null) synchronized {
+      if (ctx.channel.attr(SESSION_HANDLE).get == null) {
+        val sessionHandle = openSession(ctx)
+        ctx.channel.attr(SESSION_HANDLE).set(sessionHandle)
+        val connectionId = ctx.channel.attr(CONNECTION_ID).get
+        connIdToSessHandle.put(connectionId, sessionHandle)
+      }
+    }
+
+  def openSession(ctx: ChannelHandlerContext): SessionHandle = synchronized {
+    try {
+      val user = ctx.channel.attr(USER).get
+      val remoteIp = ctx.channel.attr(REMOTE_IP).get
+      // TODO parse SET command, save other variables at ChannelHandlerContext
```

Comment:
mysql text protocol does not support pass session confs and vars when open connection